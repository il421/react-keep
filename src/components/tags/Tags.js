import React, { Component, createRef } from 'react';
import {connect} from 'react-redux';

import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import Modal from 'react-modal';

import { handleAddTag, handleRemoveTag, handleUpdateTag } from '../../actions/tags';
import { removeTagFromNotes, updateNotesTag } from '../../actions/notes';

import TagsList from './TagsList';

export class Tags extends Component{
  constructor(props) {
    super(props);
    this.state = {
      displayEditTags: false,
      editableTag: null,
      tag: '',
    };

    // input new tag ref
    this.refNewInput = createRef();
  }

  toggleEditTags = () => {
    this.setState((prevState) => ({
      displayEditTags: !prevState.displayEditTags
    }));
    this.props.hideSidebar();
  }

  onTagChange = (evt) => {
    const value = evt.target.value;
    this.setState(() => ({
      tag: value.trim()
    }));
  }

  onCurrentTagUpdate = (id, update) => {
    if (update.trim().length > 0) {
      this.props.handleUpdateTag(id, update);
      this.props.updateNotesTag(id, update);
    }

    this.setEditableTag(id);
  }

  addNewTag = () => {
    if (this.state.tag !== '') {
      this.props.handleAddTag(this.state.tag);
    } else {
      // Add toast
    }

    // clean state.tag and local input value by ref
    this.refNewInput.current.value = '';
    this.setState({ tag: '' });
  }

  removeTag = (id) => {
    this.props.handleRemoveTag(id);
    this.props.removeTagFromNotes(id);
  }

  setEditableTag = (id) => {
    let editableTag = null;

    if (this.state.editableTag === null) {
      editableTag = id;
    }

    this.setState({ editableTag });
  }

  render() {
    return (
      <div className="tags">

        <div className="tags__filters">
          <TagsList userTags={ this.props.tags } name={ 'list' } />
        </div>

        <div className="tags__add-edit pointer">
          <FontAwesomeIcon icon="edit" size="2x" />
          <button onClick={ this.toggleEditTags } className="button--link">Add/Edit Tags</button>
        </div>

        <Modal
          isOpen={ this.state.displayEditTags }
          onRequestClose={ this.toggleEditTags }
          closeTimeoutMS={ 200 }
          className="tags__modal"
          ariaHideApp={ false }
        >
          <div>
            <div className="tags__title">Add/Edit Tags</div>
            <div className="tags__add">
              <input
                id="newTag"
                ref={ this.refNewInput }
                type="text"
                placeholder="Type a new tag ..."
                onChange={ this.onTagChange }
                onKeyPress={ (evt) => evt.key === 'Enter' && this.addNewTag() }
                disabled={ this.state.editableTag !== null }
              />
              <button className="button--link" onClick={ this.addNewTag }>+</button>
            </div>

            <div className="tags__list">
              {
                this.props.tags.map((tag, index) => (
                  <div className="tags__item item" key={ index }>

                    {
                      this.state.editableTag === tag.id ?
                      // if id is in editableTags array display input to edit the tag
                        (
                          <>
                            <input
                              id="editableTag"
                              type="text"
                              autoFocus={ true }
                              onChange={ this.onTagChange }
                              onKeyPress={ (evt) => evt.key === 'Enter' && this.onCurrentTagUpdate(tag.id, this.state.tag) }
                              placeholder={ tag.value }
                            />
                            <div className="item__actions">
                              <div className="pointer" onClick={ () => this.onCurrentTagUpdate(tag.id, this.state.tag) }>
                                <FontAwesomeIcon icon="edit" size="1x" />
                              </div>
                              <div className="pointer" onClick={ () => this.setEditableTag(tag.id) }>
                                <FontAwesomeIcon icon="times" size="1x" />
                              </div>
                            </div>
                          </>
                        ) : (
                          <>
                            <div className="pointer" onClick={ () => this.setEditableTag(tag.id) }>
                              <FontAwesomeIcon icon="edit" size="1x" />
                            </div>
                            <div className="tags__name">{ tag.value }</div>
                            <div className="pointer" onClick={ () => {this.removeTag(tag.id)} }>
                              <FontAwesomeIcon icon="times" size="1x" />
                            </div>
                          </>
                        )
                    }
                  </div>
                ))
              }
            </div>
          </div>
        </Modal>
      </div>
    );
  }
}

const mapDispatchToProps = (dispatch) => ({
  handleAddTag: (tag) => dispatch(handleAddTag(tag)),
  handleRemoveTag: (id) => dispatch(handleRemoveTag(id)),
  removeTagFromNotes: (id) => dispatch(removeTagFromNotes(id)),
  handleUpdateTag: (id, update) => dispatch(handleUpdateTag(id, update)),
  updateNotesTag: (id, update) => dispatch(updateNotesTag(id, update))
}
);

const mapStateToProps = (state) => {
  return {
    tags: state.tags
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(Tags);
