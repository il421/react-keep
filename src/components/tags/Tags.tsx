import React from "react";
import "../../styles/components/tags/_tags.scss";
import TagForm from "./TagForm";
import TagsList from "./TagsList";

interface TagsProps {}

class Tags extends React.PureComponent<TagsProps> {
  render() {
    return (
      <div className="tags">
        <TagForm />
        <TagsList />
      </div>
    );
  }
}

export default Tags;
