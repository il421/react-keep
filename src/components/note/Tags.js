import React from 'react';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import Modal from 'react-modal';
import { startAddTag, startRemoveTag } from '../../actions/tags';
import {connect} from 'react-redux';

export class Tags extends React.Component{
  constructor(props) {
    super(props);
    this.state = {
      displayEditTags: false,
      tag: '',
    };
  }

  toggleEditTags = () => {
    this.props.toggleSidebarDisplay();

    this.setState((prevState) => ({
      displayEditTags: !prevState.displayEditTags
    }));
  }

  onTagChange = (evt) => {
    const value = evt.target.value;
    this.setState(() => ({
      tag: value.trim()
    }));
  }

  addNewTag = () => {
    if (this.state.tag !== '') {
      this.props.startAddTag(this.state.tag);
    } else {
      // Add toast
    }

    this.setState({ tag: '' });
  }

  render() {
    return (
      <div className="tags">
        <div className="tags__filters"></div>
        <div className="tags__add-edit pointer">
          <FontAwesomeIcon icon="edit" size="2x" />
          <button onClick={ this.toggleEditTags } className="button--link">Add/Edit Tags</button>
        </div>

        <Modal
          isOpen={ this.state.displayEditTags }
          onRequestClose={ this.toggleEditTags }
          contentLabel="Add/Edit Tags"
          closeTimeoutMS={ 200 }
          className="tags__modal"
          ariaHideApp={ false }
          overlayClassName="modal-overlay"
        >
          <div>
            <div className="tags__add">
              <input
                type="text"
                placeholder="Type a new tag ..."
                autoFocus={ true }
                onChange={ this.onTagChange }
                value={ this.state.tag }
              />
              <button className="button--link" onClick={ this.addNewTag }>+</button>
            </div>

            <div className="tags__list">
              {
                this.props.tags.length > 0 &&
                this.props.tags.map((tag, index) => (
                  <div key={ index } className="tags__item">
                    <div><FontAwesomeIcon icon="edit" size="1x" /></div>
                    <div>{ tag.value }</div>
                    <div className="pointer" onClick={ () => this.props.startRemoveTag(tag.id) }>
                      <FontAwesomeIcon icon="times" size="1x" />
                    </div>
                  </div>
                ))
              }
            </div>
          </div>
        </Modal>
      </div>
    );
  }
}

const mapDispatchToProps = (dispatch) => ({
  startAddTag: (tag) => dispatch(startAddTag(tag)),
  startRemoveTag: (id) => dispatch(startRemoveTag(id))
}
);

const mapStateToProps = (state) => {
  return {
    tags: state.tags
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(Tags);
