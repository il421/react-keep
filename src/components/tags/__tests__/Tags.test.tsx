import React from "react";
import { shallow, ShallowWrapper } from "enzyme";
import Tags from "../Tags";
import toJson from "enzyme-to-json";

let wrapper: ShallowWrapper;

test("should render Tags", () => {
  wrapper = shallow(<Tags />);
  expect(toJson(wrapper)).toMatchSnapshot();
});
