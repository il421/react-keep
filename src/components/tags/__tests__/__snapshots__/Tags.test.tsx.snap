// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Tags 1`] = `
<Fragment>
  <Connect(TagForm) />
  <Connect(TagsListBase) />
</Fragment>
`;
